### ISSUE TITLE
【IOS WeChat 】No return data!

### ISSUE THREAD

**Issue by FeSeason** [fingerprintjs/fingerprintjs#645]
### Scenario

userAgent: `
Mozilla/5.0 (iPhone; CPU iPhone OS 14_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 MicroMessenger/8.0.2(0x18000239) NetType/WIFI Language/zh_CN
`

- use cdn last version sdk `fingerprintjs@3`

```
    <script>
      function initFingerprintJS() {

        // Initialize an agent at application startup.
        const fpPromise = FingerprintJS.load();

        // Get the visitor identifier when you need it.
        fpPromise
          .then((fp) => {
             // Here can get  fp
             return  fp.get()
           })
          .then((result) => {
            // No executed!! 

            // This is the visitor identifier:
            const visitorId = result.visitorId;
            alert(visitorId);
          });
      }
    </script>
    <script
      async
      src="//cdn.jsdelivr.net/npm/@fingerprintjs/fingerprintjs@3/dist/fp.min.js"
      onload="initFingerprintJS()"
    ></script>

```

**Comment by Finesse** [comment]
@FeSeason Thank you for the report. I'm not familiar with WeChat, so could you please give steps to reproduce?

**Comment by FeSeason** [comment]
AppLink【ios】: https://apps.apple.com/cn/app/wei/id414478124

- Step 1, 
Copy the official case code and start local server 

- Step 2,
Use WeChat to scan or open link in WeChat

- Step 3，
Problem found ..

**Comment by yuhonyon** [comment]
同样遇到  ios wechat 无返回

**Comment by caohongtao** [comment]
> AppLink【ios】: https://apps.apple.com/cn/app/wei/id414478124
> 
> * Step 1,
>   Copy the official case code and start local server
> * Step 2,
>   Use WeChat to scan or open link in WeChat
> * Step 3，
>   Problem found ..
> 
> @Finesse

same problem in IOS WeChat

**Comment by Finesse** [comment]
@FeSeason @yuhonyon @caohongtao I need your help. Please scan this QR with WeChat so that I can sign up and debug the issue:

![image](https://user-images.githubusercontent.com/9006227/116172352-f7fff800-a74d-11eb-8b48-911a429d704d.jpeg)

**Comment by Finesse** [comment]
Please give me access to WeChat or investigate the problem by yourself, otherwise I don't be able to help you

**Comment by Finesse** [comment]
Please give me access to WeChat or investigate the problem by yourself, otherwise I don't be able to help you. Please try to use [NPM installation method](https://github.com/fingerprintjs/fingerprintjs#alternatively-you-can-install-from-npm-to-use-with-webpackrollupbrowserify) and let me know if it helps.

**Comment by Perpherior** [comment]
Hi @Finesse, tried to scan the code but it says QR code is expired. Could you share a new one?

**Comment by FeSeason** [comment]
@Finesse @Perpherior 
![image](https://user-images.githubusercontent.com/12840843/116347912-af187400-a81f-11eb-92b5-565458675bb0.png)

If registration is difficult,

You can give a link that I open it on wechat, you can debug remotely?

**Comment by Finesse** [comment]
@Perpherior @FeSeason Lets try registration first. Please try this code:

![D0E634A7-060F-4389-8640-18DA08F2C377](https://user-images.githubusercontent.com/9006227/116348324-420dda00-a831-11eb-99ae-4d613a9f6e7c.jpeg)

If I can’t register, then we will try to debug remotely.

**Comment by FeSeason** [comment]
> @Perpherior @FeSeason Lets try registration first. Please try this code:
> 
> ![D0E634A7-060F-4389-8640-18DA08F2C377](https://user-images.githubusercontent.com/9006227/116348324-420dda00-a831-11eb-99ae-4d613a9f6e7c.jpeg)
> 
> If I can’t register, then we will try to debug remotely.

Done!

**Comment by Finesse** [comment]
@FeSeason Thank you! I’ll check the bug soon

**Comment by Finesse** [comment]
I've found the reason of the bug. The [fontPreferences entropy source](https://github.com/fingerprintjs/fingerprintjs/blob/d2a1401b2724fdf548ce0315e8f9d820e219d575/src/sources/font_preferences.ts) doesn't complete. It [creates an iframe](https://github.com/fingerprintjs/fingerprintjs/blob/d2a1401b2724fdf548ce0315e8f9d820e219d575/src/utils/dom.ts#L29) with an [srcdoc](https://developer.mozilla.org/en-US/docs/Web/API/HTMLIFrameElement/srcdoc) attribute and waits for the iframe to load, but the iframe doesn't load in WeChat browser for some reason.

I've found a solution. If I move the `iframe.srcdoc = initialHtml` line below the `document.body.appendChild(iframe)` line, the iframe starts getting loaded and `get()` returns a result. I'll publish a fix soon.

**Comment by Finesse** [comment]
The fix will be published in v3.1.1

**Comment by zerrol** [comment]
This problem was fixed in ios 14.x，but the problem still exisits in wechat with **ios 13.x.** 
Please using wechat with ios 13.x . See  this url: https://dev.hunliji.com/p/frontend/fingerprint-test/index.html. 
You can see ` alert(visitorId);` did not excute

code in this html :
``` html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  hello
  <script>
    function initFingerprintJS() {

      // Initialize an agent at application startup.
      const fpPromise = FingerprintJS.load();

      alert('hello 0')
      // Get the visitor identifier when you need it.
      fpPromise
        .then((fp) => {
           // Here can get  fp
           alert('hello 1')
           return  fp.get()
         })
        .then((result) => {
          // No executed!!

          // This is the visitor identifier:
          const visitorId = result.visitorId;
          alert(visitorId);
        });
    }
  </script>
  <script
    async
    src="https://cdn.jsdelivr.net/npm/@fingerprintjs/fingerprintjs@3.1.1/dist/fp.min.js"
    onload="initFingerprintJS()"
  ></script>

</body>
</html>

```

**Comment by zerrol** [comment]
@Finesse

**Comment by Finesse** [comment]
Considering iOS 13. The [fontPreferences](https://github.com/fingerprintjs/fingerprintjs/blob/edb9cbb5e61f78394d46637a5147537f441007ed/src/sources/font_preferences.ts) and the [fonts](https://github.com/fingerprintjs/fingerprintjs/blob/edb9cbb5e61f78394d46637a5147537f441007ed/src/sources/fonts.ts) entropy sources don't complete. They [create iframes](https://github.com/fingerprintjs/fingerprintjs/blob/edb9cbb5e61f78394d46637a5147537f441007ed/src/utils/dom.ts#L29) and wait for the iframes to load, but the iframes don't load in WeChat browser for some reason. I've tried using both `src` and `srcdoc`, have tried changing the command order, nothing helps. Setting a real URL to the iframe `src` attribute helps, but making network requests isn't an option.

The solution that I've found is waiting until `iframe.contentWindow.document.readyState` becomes `'complete'` (besides waiting for `onload` to fire). It fixes the problem in WeChat, keeps the correct iframe environment in Android Firefox and makes the source run faster in other browsers.

@zerrol Fixed by e11ed73055892a1d0d939e12e9fce70a0ae0ce7d. The fixed will be published in the next library release. If you need the fix soon, you can use the current master version: [fp.min.js.zip](https://github.com/fingerprintjs/fingerprintjs/files/6463016/fp.min.js.zip).

**Comment by Finesse** [comment]
In fact, the problem happened in any WKWebView, not only in WeChat. The findings above are suitable for any application that uses WKWebView.

**Comment by Finesse** [comment]
The problem still happens in iOS 12.

@jyxu1016 Unfortunately, I don't have an iOS 12 device to debug, so you need to debug the issue by yourself. A probable reason is that [an iframe loading promise](https://github.com/fingerprintjs/fingerprintjs/blob/71f2123c8e61dc00323632de3777d17c47349b38/src/utils/dom.ts#L29) never settles, therefore the `withIframe` function and the `fonts` and `fontPreferences` components never complete. You can check whether it's true on your device and try to find a way to know when the iframe is loaded to resolve the promise.

You can start your FingerprintJS development playground to debug the issue (download the repository's code and follow [the instructions](https://github.com/fingerprintjs/fingerprintjs/blob/master/contributing.md#development-playground)). Once you start a playground and open it in WeChat, add `alert()`s to the source to see what's going on. 

I'll try to research the bug in a Cordova application on iOS 12 simulator, but it won't be a 100% reliable solution.

**Comment by jyxu1016** [comment]
the problem also exisits in wechat with ios 13.x， read the comments above,  this problem has always existed， IOS wechat is not available

**Comment by Finesse** [comment]
@jyxu1016 I can't reproduce it on iOS 13.6.1 WeChat 8.0.6. I open https://fingerprintjs.github.io/fingerprintjs and see this:

![IMG_844083B7A8CB-1](https://user-images.githubusercontent.com/9006227/120291164-5c424880-c306-11eb-8437-84e2ae69d4cf.jpeg)

**Comment by jyxu1016** [comment]
I open https://fingerprintjs.github.io/fingerprintjs/ ，I can get it, too

But using the following is not executed

fpPromise
  .then((fp) => {
     // Here can get  fp
     alert('hello 1')
     return  fp.get()
   })
  .then((result) => {
    // ----------》No executed!! 《-------------
    const visitorId = result.visitorId;
    alert(visitorId);
  });

**Comment by Finesse** [comment]
@jyxu1016 I've realized that the [fix](https://github.com/fingerprintjs/fingerprintjs/commit/e11ed73055892a1d0d939e12e9fce70a0ae0ce7d) for iOS 13 hasn't been published yet, sorry. If you need the fix now, you can use the current master version: [fp.min.js.zip](https://github.com/fingerprintjs/fingerprintjs/files/6463016/fp.min.js.zip). Please subscribe to the repository updates to get notified when the fix is published to NPM.

**Comment by jyxu1016** [comment]
https://fingerprintjs.github.io/fingerprintjs/：it is available in WeChat on iOS 12 
 hope to published as soon as possible, thank

**Comment by Finesse** [comment]
@jyxu1016 Thank you. I think we will publish it soon.

**Comment by jyxu1016** [comment]
I'm waiting for you to publish every day

**Comment by jyxu1016** [comment]
@Finesse  I'm waiting for you to publish every day

**Comment by Finesse** [comment]
@jyxu1016 The fix has just been released in [v3.1.3](https://github.com/fingerprintjs/fingerprintjs/releases/tag/v3.1.3)
